runtime/stdlib.js

```diff
@@ -70,14 +94,40 @@ function caml_call_gen(f, args) {
   } else {
     argsLen--;
     var k = args [argsLen];
-    return k (function () {
-      var extra_args = (arguments.length == 0)?1:arguments.length;
-      var nargs = new Array(argsLen + extra_args);
-      for(var i = 0; i < argsLen; i++ ) nargs[i] = args[i];
-      for(var i = 0; i < arguments.length; i++ )
-        nargs[argsLen + i] = arguments[i];
-      return caml_call_gen(f, nargs)
-    });
+    switch (d) {
+    case 1: {
+      var g = function (x, y){
+        var nargs = new Array(argsLen + 2);
+        for(var i = 0; i < argsLen; i++ ) nargs[i] = args[i];
+        nargs[argsLen] = x;
+        nargs[argsLen + 1] = y;
+        return f.apply(null, nargs)
+      };
+      break;
+    }
+    case 2: {
+      var g = function (x, y, z){
+        var nargs = new Array(argsLen + 3);
+        for(var i = 0; i < argsLen; i++ ) nargs[i] = args[i];
+        nargs[argsLen] = x;
+        nargs[argsLen + 1] = y;
+        nargs[argsLen + 2] = z;
+        return f.apply(null, nargs)
+      };
+      break;
+    }
+    default: {
+      var g = function (){
+        var extra_args = (arguments.length == 0)?1:arguments.length;
+        var nargs = new Array(argsLen + extra_args);
+        for(var i = 0; i < argsLen; i++ ) nargs[i] = args[i];
+        for(var i = 0; i < arguments.length; i++ )
+          nargs[argsLen + i] = arguments[i];
+        return caml_call_gen(f, nargs)
+      };
+    }}
+    g.l = d;
```

Comment:
the functions `g` defined above have arity 2 when d = 1, and arity 3 when d = 2

Reply:
Indeed! Thanks!